walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,117 @@
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
 
 public final class FindMeetingQuery {
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    int requestDuration = (int) request.getDuration();
+
+    //if there are no events and the request duration is less than a day, return the whole day
+    if (events.isEmpty() && requestDuration <= TimeRange.WHOLE_DAY.duration()){
+        return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+    //if the requested meeting is longer than a day, return an empty list since this is not possible
+    if (requestDuration > TimeRange.WHOLE_DAY.duration()) {
+        return Collections.emptyList();
+    }
+
+    //an arraylist to hold an ordered list without duplicates of all busy times in a day
+    ArrayList<TimeRange> busyTimes = new ArrayList<TimeRange>();
+
+    //adding the beginning and end of the day as endpoints (your meeting can't be before the beginning of the day or after the end of the day)
+    busyTimes.add(TimeRange.fromStartDuration(0, 0));
+    busyTimes.add(TimeRange.fromStartDuration(1440,0));
+
+    for (Event event : events){
+        //create a timerange for the event we're evaluating
+        TimeRange newEvent = TimeRange.fromStartEnd(event.getWhen().start(), event.getWhen().duration()+event.getWhen().start(), false);
+        int newEventStart = newEvent.start();
+        int newEventEnd = newEvent.end();
+
+        //if none of the attendees in the request are attending this event, don't add it to BusyTimes because it's not relevant
+        Set<String> eventAttendees = event.getAttendees();
+        Collection<String> requestAttendees = request.getAttendees();
+        if (Collections.disjoint(eventAttendees, requestAttendees)){
+            continue;
+        }
+
+        for (int i = 0; i < busyTimes.size()-1; i++) {
+            int currentBusyStart = busyTimes.get(i).start();
+            int currentBusyEnd = busyTimes.get(i).end();
+            int nextBusyStart = busyTimes.get(i+1).start();
+            int nextBusyEnd = busyTimes.get(i+1).end();
+            int adjustedNewStart;
+            int adjustedNewEnd;
+
+            //if the current busy time contains the new event, or they are equal, do not add the new event to the arraylist as it already exists
+            /*****/  //current busy block
+             /**/    //new event
+            if (busyTimes.get(i).contains(newEvent) || busyTimes.get(i).equals(newEvent)){
+                break; 
+            }
+            //if the new event starts during the current busy time and ends before the next busy time
+            /*******/        /**/       //busy blocks
+                 /******/               //new event
+            else if (TimeRange.contains(busyTimes.get(i), newEventStart) && newEventEnd <= nextBusyStart) {
```

Comment:
This overall logic is complicated, hard to follow and (I think) missing some edge cases (e.g. what if the new event overlaps with multiple events). I think a better alternative would be to
1. find ALL events that overlap with the new event
2. replace them with new event that for which startTime = min(all affected events start time) and endTime = max(all affected events end time)